mono/metadata/appdomain.h

```diff
@@ -30,12 +30,21 @@ typedef void (*MonoDomainFunc) (MonoDomain *domain, void* user_data);
 MONO_API MonoDomain*
 mono_init                  (const char *filename);
 
+MONO_API MonoDomain*
+mono_init_w                (const wchar_t *filename);
```

Comment:
What is the purpose of this?  right now the mono API takes UTF8-encoded `char*` arguments.  If an application has some other sort of text, it should convert it itself before calling into Mono.

This will add a lot of exported functions increasing the size of libmono even for API consumers that will never use it (e.g. the Xamarin products)

Reply:
The increase is small. Surely there are no other encodings. See the bug.

Reply:
I handled the implementation-defined size.

Reply:
You know -- char is implemention defined, and there really are multiple encodings. But we still use it and assume it is utf8, even though it isn't always.

Reply:
Java code and C# traffics heavily in 2 byte char. Mono therefore traffics heavily in 2 byte char. A little bit of wchar_t is not a terrible thing, and could actually help customers. It is true this customer is confusing encoding of char. This change would help steer them.

Reply:
There is no need to introduce `wchar_t` here. Please use `char`, `gunichar` or `gunichar2`. On top of that, we have no interest in introducing new public API expect if absolutely necessary, and it's definitely not the case here.

Reply:
Char unichar & unichar2 are deliberately not used here. Wchar_t is a heavily used type that should be usable. It does resolve to either unichar or unichar2 but that resolution is trivially within the implementation.

Reply:
I think we should add api which is actually used/requested by users.

Reply:
This users code would benefit from some of this, though they didn't know to ask. I added a bit more than their code directly implies. One function would suffice. The rest starts to fill in the obvious other pieces -- other than directly requested, there is also what holds together as complete. Anyone with non-English paths would benefit esp. on Windows such as this user.

Reply:
We do support `wchar_t` already:  If you have a wide character string, convert it to UTF-8 and call one of our existing APIs.  

Also I don't think we should be doing our own ad-hoc charset conversion instead of using a portable well-tested library.

Reply:
Yeah I was thinking about that too, but our library maybe isn't terrible and we use it heavily --  well, some of it. We use utf8 and ucs2 a lot, but maybe not ucs4. Non-portable there is Windows MultiByteToWideChar and WideCharToMultiByte. Otherwise ICU is a heavily used library, and was recently added to Windows. This stuff isn't that complicated that handling it yourself is so bad. This user doesn't need ucs4. We could ifdef windows and use MultiByteToWideChar/WideCharToMultiByte but consistency of doing it ourselves the same way on all platforms might be better.

Reply:
Aha wcstombs_l/xlocale at least on some systems e.g. Mac and sometimes Linux.